compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -923,14 +923,22 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                         [self_ty, Ty::new_tup(tcx, sig.inputs())],
                     )
                 });
+
                 // We must additionally check that the return type impls `Future`.
+
+                // FIXME(async_closures): Investigate this before stabilization.
+                // We instantiate this binder eagerly because the `confirm_future_candidate`
+                // method doesn't support higher-ranked futures, which the `AsyncFn`
+                // traits expressly allow the user to write. To fix this correctly,
+                // we'd need to instantiate trait bounds before we get to selection,
+                // like the new trait solver does.
                 let future_trait_def_id = tcx.require_lang_item(LangItem::Future, None);
+                let placeholder_output_ty = self.infcx.enter_forall_and_leak_universe(sig.output());
```

Comment:
so, if actually used (via unboxed closures), we'd just get odd errors about traits not being implemented even though they are?

Reply:
No, this change only affects `async Fn` trait goals, and makes them go from ICE -> PASS.

That issue I linked via unboxed closures still ICEs; I'll put up a fix like #108834 in a separate PR.